lib/phoenix/channel/server.ex

```diff
@@ -0,0 +1,115 @@
+defmodule Phoenix.Channel.Server do
+  use GenServer
+  require Logger
+  alias Phoenix.PubSub
+
+  @moduledoc """
+  Handles `%Phoenix.Socket{}` state and invokes channel callbacks.
+
+  ## handle_info/2
+  Regular Elixir messages are forwarded to the socket channel's
+  `handle_info/2` callback.
+
+  ## exits
+  Exists are trapped by default and the socket channel's `leave/2`
+  callback is triggered.
+  """
+
+  def start_link(socket, auth_payload) do
+    GenServer.start_link(__MODULE__, [socket, auth_payload])
+  end
+
+  def init([socket, auth_payload]) do
+    Process.flag(:trap_exit, true)
+
+    case socket.channel.join(socket.topic, auth_payload, socket) do
+      {:ok, socket} ->
+        {:ok, socket}
+          socket = put_in(socket, [:pid], self)
+          Process.link(socket.adapter_pid)
```

Comment:
You don't need this link because the adapter already calls start_link. Unless the adapter_pid can be different of the started_pid. Is that the case?